### ISSUE TITLE
Default export TypeScript compatibility issue

### ISSUE THREAD

**Issue by jjrv** [systemjs/systemjs#667]
In TypeScript 1.5.0 the new ES6 `export` syntax:

```
export default class Foo {};
```

compiles to:

```JavaScript
var Foo = (function () {
    function Foo() {
    }
    return Foo;
})();
exports.default = Foo;
```
But System.js copies all members of the `exports` object into a new object, and finally sets its `default` property to the original exports object ([see code](https://github.com/systemjs/systemjs/blob/de2eddd566f6026df972f43d7de1c6a2e3dda061/lib/register.js#L410)) - which already had a `default` property!

As a result, using the imported module fails because TypeScript refers to `Foo.default` while the class constructor is `Foo.default.default`

I'd suggest skipping the offending line if the `exports` object already contains a `default` property.

**Comment by guybedford** [comment]
Rather use the `system` module format output in TypeScript.

**Comment by jjrv** [comment]
The system module output format is the one that specifically fails. It uses System.register to load other modules and if those other modules are in CommonJS format (as distributed precompiled in another project) then SystemJS fails to load them.

**Comment by jjrv** [comment]
If I understand correctly, the offending line is a hack to support older code which uses the entire `exports` object as exports.default should be used. That hack then breaks newer code, which already places the default export in the correct place. I think the hack should be conditional.

**Comment by jjrv** [comment]
Also, here's an example of another project correctly handling the issue (https://github.com/calvinmetcalf/es6-translate):
> Inside CommonJS modules, if there is a key named default in a module it is assumed to be the default export and just that is imported

([relevant code](https://github.com/calvinmetcalf/es6-translate/blob/master/lib/evaluate.js#L40))

**Comment by guybedford** [comment]
@jjrv that won't work in general, as there are CommonJS modules that exist with default exports already, and we can't break them.

**Comment by jjrv** [comment]
@guybedford why would it break them? If the module defines `exports.default`, it's used as is. Otherwise the entire `exports` becomes the default export. A configuration option also allows turning the autodetection off.

As far as I can see, that feature would be the only way that a TypeScript module compiled for SystemJS could load a module compiled for CommonJS with a default export. I tried to fix the issue with a custom loader plugin, but the exports.default hack happens after the `instantiated` hook so even that doesn't help. The only (horribly ugly) fix would be to copy members of `exports.default` to `exports` in a custom loader plugin, so SystemJS then copies them back, merging all exports into `exports.default`.

Would you accept a PR where the fix is toggled with a configuration option? I'd really like to use SystemJS and CommonJS modules together and it seems there should be an option to do so.

**Comment by guybedford** [comment]
Joi is an example library that has an `exports.default` in https://github.com/hapijs/joi/blob/master/lib/any.js#L347. This was the example that broke when Babel took this interop path.

**Comment by guybedford** [comment]
Just let me know if you have any further questions here.

**Comment by jjrv** [comment]
In case someone finds this issue later, it's further discussed in TypeScript issue 5285, apparently with a fix coming in TypeScript 1.8.
